### PR TITLE
feat(joins) Support composite queries in db_query step 1: Add a visitor for data sources.

### PR DESCRIPTION
**Pull request by fpacifici**
I noticed we are running into quite some cases where we need to process the tree of data sources in a composite query.
One of the cases was formatting.

Initially I could not add a visitor because the data sources were not defined in the same module. That is still true, though we can do something similar by encapsulating the exhaustive `isinstance` checks in a class that gets extended, so this is not spread around the code base and we can still rely on a visitor style abstraction.

### CODE REVIEW

File: snuba/clickhouse/formatter/query.py [hunk 1]
```diff
@@ -48,39 +51,37 @@ def format_query(query: FormattableQuery, settings: RequestSettings) -> Formatte
...
+        """
 
-def format_table(data_source: Table) -> StringNode:
-    """
-    Formats a simple table data source.
-    """
+        final = " FINAL" if data_source.final else ""
+        sample = (
+            f" SAMPLE {data_source.sampling_rate}" if data_source.sampling_rate else ""
+        )
```
Comment (lynnagara) [inline snuba/clickhouse/formatter/query.py:69]: We may have discussed this but remind me again - are final and sample just getting applied to each one of the referenced data sources independently?

**Review by lynnagara — APPROVED** [review]

**Review by evanh — APPROVED** [review]

File: snuba/clickhouse/formatter/query.py (hunk 1 above)
Comment (fpacifici) [reply snuba/clickhouse/formatter/query.py:69]: Yes they are. As per clickhouse docs FINAL and SAMPLE are property of the subquery. https://clickhouse.tech/docs/en/sql-reference/statements/select/from/#select-from

Comment (lynnagara) [reply snuba/clickhouse/formatter/query.py:69]: I mean for us sample and final are in the main query, so are we just going to push it down to every subquery equally?

Comment (fpacifici) [reply snuba/clickhouse/formatter/query.py:69]: They are associated with each entity in SnQL they are not present in the main query. The old language does not support composite queries.

**Pull request merged**